Choose JVM options ergonomically

danielmitterdorfer opened the pull request:
With this commit we add the possibility to define further JVM options
(and system properties) based on the current environment. As a proof of
concept, it chooses Netty's allocator ergonomically based on the maximum
defined heap size.

This PR is work in progress and just up for the curious. Especially, we need to
run experiments to decide at which heap size to switch the allocator from
pooled to unpooled.

danielmitterdorfer committed:
Choose JVM options ergonomically

With this commit we add the possibility to define further JVM options
(and system properties) based on the current environment. As a proof of
concept, it chooses Netty's allocator ergonomically based on the maximum
defined heap size.

nik9000 reviewed (commented):
I think it is pretty cute and it could be quite useful down the road.

I expect it'd be easier to get the properties if we started the launcher with the user defined jvm arguments but I expect that could cause funny issues down the road. Also it'd be silly to start the launcher with a huge heap. So I think your approach here is probably better than starting the launcher with the user defined jvm arguments. Though it is nice to think about it.

s1monw reviewed (approved):
LGTM

dakrone commented on distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/JvmErgonomics.java:
```diff
@@ -0,0 +1,106 @@
+                }
+            }
+        }
+        return null;
```
Would it be possible to return what the default JVM heap is here instead of null? Seems like we still may want to make changes to JVM options for an unset heap.

Either that, or since we explicitly set it in our `jvm.options`, emit a warning or error that no heap has been specified (someone must have removed the option)

jasontedor replied:
The JVM chooses the heap size ergonomically when the heap size is not specified.

dakrone replied:
> The JVM chooses the heap size ergonomically when the heap size is not specified.

Do we have the ability to see what it would choose? If running on a small machine, for instance, we'd still want to disable Netty's pooled allocator if the JVM is going to automatically choose a 400mb heap

jasontedor replied:
It can be quite tricky because the ergonomics these days depend on other flags on the JVM (e.g., `-XX:+UseCGroupMemoryLimitForHeap`).

jasontedor replied:
We could try starting a JVM with all the flags specified and `-XX:+PrintFlagsFinal -version` and scrape the output for `MaxHeapSize`.

danielmitterdorfer replied:
Thanks for the correction. I've corrected my comment accordingly.

jasontedor replied:
> We don't remove the options and use `-XX:+UseCGroupMemoryLimitForHeap` for our docker image(s) do we?

We do not do that.

danielmitterdorfer committed:
Merge remote-tracking branch 'origin/master' into ergonomics

danielmitterdorfer committed:
Switch allocator at 1GB heap and be explicit

With this commit we switch to the unpooled allocator at 1GB heap size
(value determined experimentally determined, see PR for more details).
We are also explicit about the choice of the allocator in either case.

danielmitterdorfer commented:
### Benchmark Scenarios

I ran benchmarks for the following data sets to determine at which point we should switch the allocator (click on the links below for a thorough description of the data set and an example document):

* [geonames](https://github.com/elastic/rally-tracks/tree/master/geonames)
* [geopoint](https://github.com/elastic/rally-tracks/tree/master/geopoint)
* [http_logs](https://github.com/elastic/rally-tracks/tree/master/http_logs)
* [nyc_taxis](https://github.com/elastic/rally-tracks/tree/master/nyc_taxis)
* [pmc](https://github.com/elastic/rally-tracks/tree/master/pmc)

I ran those for one node, three nodes (one replica) and one node with X-Pack Security enabled. Heap size was 512MB, 768MB and 1G. All benchmarks are based on revision 37f67d9e21b.

### Results

Raw results can be downloaded from [raw_results.zip](https://github.com/elastic/elasticsearch/files/2111016/raw_results.zip). Here I just want to highlight some of the results.

#### GC metrics

Overall old GC decreases (sometimes significantly) at the expense of slight increase of young GC. This is expected because there is no pool that takes up space (in the old generation) but rather we allocate new objects for every request most of which are short lived.

This is a rather typical case from geonames:

One node:

![gc_times](https://user-images.githubusercontent.com/1699576/41532957-5389b120-72f9-11e8-9383-66771446c9b2.png)

Three nodes, one replica:

![gc_times](https://user-images.githubusercontent.com/1699576/41532965-5b19dec4-72f9-11e8-9b5a-46c7b66da811.png)

For other benchmarks (fulltext, i.e. PMC) we are able to use lower heap sizes when we switch to the unpooled allocator:

![gc_times](https://user-images.githubusercontent.com/1699576/41532979-656b6438-72f9-11e8-9a60-4f4ad1c6c17e.png)

Note that the benchmark did not finish successfully for a 512MB heap in either case but with the unpooled allocator we were able to finish with 768MB.

#### Indexing Throughput

Throughput increases in almost all cases except when we have rather small documents (i.e. geopoints). Here are several examples (the bars show the median; black lines indicate minimum and maximum):

For small documents (one property with one geopoint per document) we see a slight decrease in indexing throughput for heap sizes > 512MB:

![throughput_index-append](https://user-images.githubusercontent.com/1699576/41532990-734d9ae4-72f9-11e8-9b86-ac46b8bf0ba3.png)

In other cases we typically see an increase in indexing throughput and less variation (i.e. the black line gets smaller).

For HTTP logs the difference gets more pronounced the less heap size we have available (one node):

![throughput_index-append](https://user-images.githubusercontent.com/1699576/41533023-94a5fefc-72f9-11e8-95c5-66946189715f.png)

geonames (one node):

![throughput_index-append](https://user-images.githubusercontent.com/1699576/41533035-9fc76c8a-72f9-11e8-822c-f3e53936578c.png)

geonames (three nodes, one replica):

![throughput_index-append](https://user-images.githubusercontent.com/1699576/41533039-a5b4690e-72f9-11e8-8a2f-dc909eb722d6.png)

For PMC the difference is quite pronounce already for 1GB heaps:

one node:

![throughput_index-append](https://user-images.githubusercontent.com/1699576/41533052-ab8b11ac-72f9-11e8-939c-84626d0fbd36.png)

Again, we were not able to finish this benchmark for 512MB heap size but succeeded for 768MB heap size with the unpooled allocator.

three nodes:

![throughput_index-append](https://user-images.githubusercontent.com/1699576/41533085-c8a72460-72f9-11e8-8791-6810880dc230.png)


#### Query Time

In almost all cases service time is roughly identical (within run-to-run variation):

This is an ideal case showing the match-all query for geonames on one Elasticsearch node:

![service_time_default](https://user-images.githubusercontent.com/1699576/41533093-d6a94e80-72f9-11e8-8301-8a23b6c7c155.png)

For scrolls we see a little bit higher variation in some cases (this is geonames with X-Pack):

![service_time_scroll](https://user-images.githubusercontent.com/1699576/41533101-dfa17026-72f9-11e8-850b-935d73d7c7ab.png)

Times also tend to be a little bit worse with the unpooled allocator at higher heap sizes (see the scroll picture above).

### Conclusion

Based on the data it is not a hard and clear decision at which point we should switch but for the most part the situation has already improved in the 1GB case. The smaller heap sizes get, the more the unpooled allocator makes sense. I'd switch at 1GB heap size to the unpooled allocator (which is already reflected in the last commit) but if we want to be more conservative we could also switch at 768MB.

@dakrone, @nik9000, @s1monw, @jasontedor are you still ok with my changes given those results? Also picking up the previous discussion on what to do when the user did not specify a heap size (i.e. `-Xms` / `_Xmx` is manually removed from `jvm.options`): I think this is quite an edge case and would ignore it for now. We can always refine later if needed.

nik9000 commented:
I'm in.

Seriously, though, you've proven to me that ergonomically picking some JVM options based on other provided options is a great idea. I'm fine with kicking this optimization in at 1GB because it looks like it mostly helps.

I wonder:
1. If a user specifies a value for an option that we'd pick ergonomically what should we do? I figure we can either accept the user's choice or refuse to start. I don't think we should overwrite their choice with ours.
2. If a user doesn't specify the heap size should we try to scrape it from running the jvm with `-XX:+PrintFlagsFinal -version`? This got me to wondering, if the user doesn't specify the heap size is it valid for us to:
2a. Refuse to start?
2b. Ergonomically pick a heap size based on the ram on the machine/container?

danielmitterdorfer commented:
Thanks for your feedback. To your thoughts:

> If a user specifies a value for an option that we'd pick ergonomically what should we do?

If a user sets some value explicitly, this effectively disables our ergonomic choice. I think this makes sense because it allows us to provide sensible defaults but still enables users to tune things on their own.

(paraphrasing you below)

> If the user doesn't specify the heap size is it valid for us to refuse to start or ergonomically pick a heap size [...]?

IMHO apart from the choice of the garbage collector, the heap size is the single most important choice that you can make in the JVM configuration. For a small tool it might be ok to not specify the heap size but for server software like Elasticsearch I don't think it is a good idea to leave the heap size unspecified. Why?

* It makes the configuration less predictable: If you deploy to machines with different configurations you might end up with different heap sizes, i.e. it is one more thing to be aware about if you don't set it explicitly.
* It allows for the minimum and maximum heap size to be different so the JVM might not allocate all heap up-front so pay the cost for potential resizing at runtime. That may or may not be acceptable depending on the user case. Also, it can happen that at the time the JVM wants to resize the heap you don't get enough memory from the OS (depending on OS configuration).

So *if* we want to change behavior I'd tend to refuse to start rather than picking something for the user. If we intend to do this, I think we should do this in a separate PR though?

nik9000 commented:
> If we intend to do this, I think we should do this in a separate PR though?

Right! But thinking about this informs my thinking about how we should handle the other ergonomic choices when the user doesn't specify the heap size. If we don't plan to start if the user doesn't specify the heap size then we don't have to handle it at all. Given that we plan to talk more about heap choices in the future, I think your choice not to set any ergonomic options if heap size isn't specified is fine for this PR. I think we should get this in.

danielmitterdorfer commented:
Thanks all for your feedback. I've merged this now to master and will check how our benchmark behave over the next few days as well.
